Markdown api error when using variable in code snippet

Hi all, 
I think there is a problem when calling gitlab4j markdown api with a code snippet inside: seems that gitlab4j api is evaluating the code, and I don't find any way to not do that.
I also tried calling directly the api thought a POST call and the result is 201 with the markdown processed.

As example (real data like host,token,project and namespace are hidden for obvious reasons):  
```
POST  https://aGitlaHost/api/v4/markdown
Content-Type: application/json
Private-Token: aToken

{"text":"Hello world! :tada: ```xml <profiles> <version>${maven-surefire-plugin.version}</version> </profiles>```", "gfm":true, "project":"aNamespace/aProjectName"}
```
the answer is: 
```
{
  "html": "<p data-sourcepos=\"1:1-1:104\" dir=\"auto\">Hello world! <gl-emoji title=\"party popper\" data-name=\"tada\" data-unicode-version=\"6.0\">🎉</gl-emoji> <code>xml &lt;profiles&gt; &lt;version&gt;${maven-surefire-plugin.version}&lt;/version&gt; &lt;/profiles&gt;</code></p>"
}
```

if I do the same text call thought gilab4j api, i get the following exception:
```
org.gitlab4j.api.GitLabApiException: The template variable 'maven-surefire-plugin.version' has no value
	at java.desktop/javax.swing.JList.fireSelectionValueChanged(JList.java:1804)
	at java.desktop/javax.swing.JList$ListSelectionHandler.valueChanged(JList.java:1818)
	at java.desktop/javax.swing.DefaultListSelectionModel.fireValueChanged(DefaultListSelectionModel.java:219)
	at java.desktop/javax.swing.DefaultListSelectionModel.fireValueChanged(DefaultListSelectionModel.java:186)
	at java.desktop/javax.swing.DefaultListSelectionModel.setValueIsAdjusting(DefaultListSelectionModel.java:723)
	at java.desktop/javax.swing.JList.setValueIsAdjusting(JList.java:2152)
	at java.desktop/javax.swing.plaf.basic.BasicListUI$Handler.mouseReleased(BasicListUI.java:2958)
	at java.desktop/java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:298)
	at java.desktop/java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:297)
	at java.desktop/java.awt.AWTEventMulticaster.mouseReleased(AWTEventMulticaster.java:297)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6654)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3345)
	at java.desktop/java.awt.Component.processEvent(Component.java:6419)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5029)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4861)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4547)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4488)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2773)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4861)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:778)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:751)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:749)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:748)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:971)
	at com.intellij.ide.IdeEventQueue.dispatchMouseEvent(IdeEventQueue.java:906)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:838)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:452)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:744)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$9(IdeEventQueue.java:451)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:802)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:505)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: org.gitlab4j.api.GitLabApiException: The template variable 'maven-surefire-plugin.version' has no value
	at org.gitlab4j.api.AbstractApi.handle(AbstractApi.java:655)
	at org.gitlab4j.api.AbstractApi.post(AbstractApi.java:343)
	at org.gitlab4j.api.MarkdownApi.getMarkdown(MarkdownApi.java:34)
Caused by: org.gitlab4j.api.GitLabApiException: The template variable 'maven-surefire-plugin.version' has no value
	... 49 more
Caused by: java.lang.IllegalStateException: The template variable 'maven-surefire-plugin.version' has no value
	at org.glassfish.jersey.client.JerseyWebTarget.getUri(JerseyWebTarget.java:111)
	at org.glassfish.jersey.client.JerseyWebTarget.request(JerseyWebTarget.java:191)
	at org.glassfish.jersey.client.JerseyWebTarget.request(JerseyWebTarget.java:36)
	at org.gitlab4j.api.GitLabApiClient.invocation(GitLabApiClient.java:796)
	at org.gitlab4j.api.GitLabApiClient.invocation(GitLabApiClient.java:761)
	at org.gitlab4j.api.GitLabApiClient.post(GitLabApiClient.java:524)
	at org.gitlab4j.api.GitLabApiClient.post(GitLabApiClient.java:494)
	at org.gitlab4j.api.AbstractApi.post(AbstractApi.java:341)
	... 51 more
Caused by: java.lang.IllegalArgumentException: The template variable 'maven-surefire-plugin.version' has no value
Caused by: java.lang.IllegalStateException: The template variable 'maven-surefire-plugin.version' has no value

	at org.glassfish.jersey.uri.UriTemplate$1ValuesFromArrayStrategy.valueFor(UriTemplate.java:996)
	at org.glassfish.jersey.uri.UriTemplate.resolveTemplate(UriTemplate.java:682)
	at org.glassfish.jersey.uri.UriTemplate.createUriComponent(UriTemplate.java:1006)
	at org.glassfish.jersey.uri.UriTemplate.createURIWithStringValues(UriTemplate.java:946)
	at org.glassfish.jersey.uri.UriTemplate.createURIWithStringValues(UriTemplate.java:882)
	at org.glassfish.jersey.uri.UriTemplate.createURI(UriTemplate.java:847)
Caused by: java.lang.IllegalArgumentException: The template variable 'maven-surefire-plugin.version' has no value

	at org.glassfish.jersey.uri.internal.JerseyUriBuilder._build(JerseyUriBuilder.java:890)
	at org.glassfish.jersey.uri.internal.JerseyUriBuilder.build(JerseyUriBuilder.java:807)
	at org.glassfish.jersey.client.JerseyWebTarget.getUri(JerseyWebTarget.java:109)
	... 58 more
```

Thanks

--- a/src/main/java/org/gitlab4j/api/MarkdownApi.java
+++ b/src/main/java/org/gitlab4j/api/MarkdownApi.java
@@ -1,9 +1,10 @@
 package org.gitlab4j.api;
 
-import javax.ws.rs.core.Form;
-import javax.ws.rs.core.Response;
 import org.gitlab4j.api.GitLabApi.ApiVersion;
 import org.gitlab4j.api.models.Markdown;
+import org.gitlab4j.api.models.MarkdownRequest;
+
+import javax.ws.rs.core.Response;
 
 /**
  * This class provides an entry point to all the GitLab API markdown calls.
@@ -30,8 +31,26 @@ public class MarkdownApi extends AbstractApi {
             throw new GitLabApiException("Api version must be v4");
         }
 
-        Form formData = new GitLabApiForm().withParam("text", text, true);
-        Response response = post(Response.Status.OK, formData.asMap(), "markdown");
+        return getMarkdown(new MarkdownRequest(text, true));
+    }
+
+    /**
+     * Render an arbitrary Markdown document.
+     *
+     * <pre><code>GitLab Endpoint: POST /api/v4/markdown</code></pre>
+     *
+     * @param markdownRequest a request of markdown transformation
+     * @return a Markdown instance with transformed info
+     * @throws GitLabApiException if any exception occurs
+     * @since GitLab 11.0
+     */
+    public Markdown getMarkdown(MarkdownRequest markdownRequest) throws GitLabApiException {
+
+        if (!isApiVersion(ApiVersion.V4)) {
+            throw new GitLabApiException("Api version must be v4");
+        }
+
+        Response response = post(Response.Status.OK, markdownRequest, "markdown");
         return (response.readEntity(Markdown.class));
     }
 }

--- a/src/main/java/org/gitlab4j/api/models/MarkdownRequest.java
+++ b/src/main/java/org/gitlab4j/api/models/MarkdownRequest.java
@@ -1,0 +1,27 @@
+package org.gitlab4j.api.models;
+
+public class MarkdownRequest {
+    private String text;
+    private boolean gfm;
+
+    public MarkdownRequest(String text, boolean gfm) {
+        this.text = text;
+        this.gfm = gfm;
+    }
+
+    public String getText() {
+        return text;
+    }
+
+    public void setText(String text) {
+        this.text = text;
+    }
+
+    public boolean isGfm() {
+        return gfm;
+    }
+
+    public void setGfm(boolean gfm) {
+        this.gfm = gfm;
+    }
+}

--- a/src/test/java/org/gitlab4j/api/TestMarkdownApi.java
+++ b/src/test/java/org/gitlab4j/api/TestMarkdownApi.java
@@ -1,0 +1,56 @@
+package org.gitlab4j.api;
+
+import org.gitlab4j.api.models.Markdown;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import static org.junit.Assert.*;
+import static org.junit.Assume.assumeTrue;
+
+@Category(IntegrationTest.class)
+public class TestMarkdownApi extends AbstractIntegrationTest {
+
+    private static final String EXPECTED_HTML_FOR_SPECIAL = "<p data-sourcepos=\"1:1-1:104\" dir=\"auto\">Hello world! <gl-emoji title=\"party popper\" data-name=\"tada\" data-unicode-version=\"6.0\">🎉</gl-emoji> <code>xml &lt;profiles&gt; &lt;version&gt;${maven-surefire-plugin.version}&lt;/version&gt; &lt;/profiles&gt;</code></p>";
+    public static final String SPECIAL_CHAR_EXAMPLE = "Hello world! :tada: ```xml <profiles> <version>${maven-surefire-plugin.version}</version> </profiles>```";
+
+    public static final String NORMAL_HTML_EXAMPLE = "<h1 data-sourcepos=\"1:1-1:4\" dir=\"auto\">\n" +
+            "<a id=\"user-content-h1\" class=\"anchor\" href=\"#h1\" aria-hidden=\"true\"></a>H1</h1>\n" +
+            "<h2 data-sourcepos=\"2:2-2:6\" dir=\"auto\">\n" +
+            "<a id=\"user-content-h2\" class=\"anchor\" href=\"#h2\" aria-hidden=\"true\"></a>H2</h2>\n" +
+            "<h3 data-sourcepos=\"3:2-3:7\" dir=\"auto\">\n" +
+            "<a id=\"user-content-h3\" class=\"anchor\" href=\"#h3\" aria-hidden=\"true\"></a>H3</h3>\n" +
+            "<h4 data-sourcepos=\"4:2-4:8\" dir=\"auto\">\n" +
+            "<a id=\"user-content-h4\" class=\"anchor\" href=\"#h4\" aria-hidden=\"true\"></a>H4</h4>\n" +
+            "<h5 data-sourcepos=\"5:2-5:9\" dir=\"auto\">\n" +
+            "<a id=\"user-content-h5\" class=\"anchor\" href=\"#h5\" aria-hidden=\"true\"></a>H5</h5>\n" +
+            "<h6 data-sourcepos=\"6:2-6:10\" dir=\"auto\">\n" +
+            "<a id=\"user-content-h6\" class=\"anchor\" href=\"#h6\" aria-hidden=\"true\"></a>H6</h6>";
+
+    private static GitLabApi gitLabApi;
+
+    @BeforeClass
+    public static void setUp() throws Exception {
+        gitLabApi = baseTestSetup();
+    }
+
+    @Before
+    public void beforeMethod() {
+        assumeTrue(gitLabApi != null);
+    }
+
+    @Test
+    public void testMarkdownWithSpecialCharacters() throws GitLabApiException {
+        Markdown markdown = gitLabApi.getMarkdownApi().getMarkdown(SPECIAL_CHAR_EXAMPLE);
+
+        assertEquals(EXPECTED_HTML_FOR_SPECIAL, markdown.getHtml());
+    }
+
+    @Test
+    public void testMarkdownWithNormalText() throws GitLabApiException {
+        Markdown markdown = gitLabApi.getMarkdownApi().getMarkdown("# H1 \n ## H2 \n ### H3 \n #### H4 \n ##### H5 \n ###### H6");
+
+        assertEquals(NORMAL_HTML_EXAMPLE, markdown.getHtml());
+    }
+}

